### PR TITLE
Apply Fixes from #1055

### PR DESCRIPTION
**Pull request by iambmelt**
Applies the changes in #1055 to a changes from `release/3.0.3` and merges to `3.0.4`

### CODE REVIEW

File: common/src/androidTest/java/com/microsoft/identity/common/CommandDispatcherTest.java [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+                    }
+                }, 0, submitLatch, submitLatch1);
+        FinalizableResultFuture<CommandResult> f = CommandDispatcher.submitSilentReturningFuture(testCommand);
+        submitLatch1.await();
+        testCommand.value = 2;
+        submitLatch.countDown();
+        testLatch.await();
+        Assert.assertTrue(f.isDone());
+        Assert.assertEquals(TEST_RESULT_STR, f.get().getResult());
+        f.isCleanedUp();
```
Comment (rpdome) [inline common/src/androidTest/java/com/microsoft/identity/common/CommandDispatcherTest.java:134]: This looks like something we can use RoboTestUtils.flushScheduler() for?

(no need to make changes in this PR)

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java [hunk 2]
```diff
@@ -73,84 +77,145 @@
...
+    private static ConcurrentMap<BaseCommand, FinalizableResultFuture<CommandResult>> sExecutingCommandMap = new ConcurrentHashMap<>();
+
+    /**
+     * Remove all keys that are the command reference from the executing command map.  Since if they key has
+     * been changed, remove will not work, construct a new map and add all keys that are not identically
+     * that key into the new map.
+     *
+     * @param command the command whose identity to use to cleanse the map.
+     */
+    private static void cleanMap(BaseCommand command) {
```
Comment (rpdome) [inline common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:92]: synchronized?

Comment (AdamBJohnsonx) [reply common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:92]: Maybe document that it must be used under the mapAccessLock?

File: common/src/androidTest/java/com/microsoft/identity/common/CommandDispatcherTest.java (hunk 1 above)
Comment (AdamBJohnsonx) [reply common/src/androidTest/java/com/microsoft/identity/common/CommandDispatcherTest.java:134]: Hah.  Yeah, I could have just gotten a reference on the executor and shut it down, I suppose.

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java (hunk 2 above)
Comment (rpdome) [reply common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:92]: Can we wrap them here? synchronize block is reentrant right?

Comment (rpdome) [reply common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:92]: I'm fine either way as soon as we can make sure the future dev are aware of this or can do the right thing)

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java [hunk 3]
```diff
@@ -73,84 +77,145 @@
...
-    private static final ConcurrentMap<BaseCommand, ResultFuture<CommandResult>> sExecutingCommandMap = new ConcurrentHashMap<>();
+
+    private static final Object mapAccessLock = new Object();
+    @GuardedBy("mapAccessLock")
+    private static ConcurrentMap<BaseCommand, FinalizableResultFuture<CommandResult>> sExecutingCommandMap = new ConcurrentHashMap<>();
+
+    /**
+     * Remove all keys that are the command reference from the executing command map.  Since if they key has
+     * been changed, remove will not work, construct a new map and add all keys that are not identically
+     * that key into the new map.
```
Comment (AdamBJohnsonx) [inline common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:88]:
```suggestion
     * that key into the new map.  <strong>MUST</strong> only be used under the mapAccessLock.
```

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java (hunk 2 above)
Comment (AdamBJohnsonx) [reply common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:92]: Yeah we could.  I'l like to dispose of this completely once we can stabilize the key space for this map either by generating stable keys or making the objects immutable.

File: common/src/androidTest/java/com/microsoft/identity/common/CommandDispatcherTest.java (hunk 1 above)
Comment (rpdome) [reply common/src/androidTest/java/com/microsoft/identity/common/CommandDispatcherTest.java:134]: could we add a TODO to migrate to that one (and get rid of the new future class) at some point in the future? (e.g. when you're working on the root cause fix)

**Review by rpdome — APPROVED** [review]

**Review by AdamBJohnsonx — APPROVED** [review]

**Pull request merged**